hail/src/main/scala/is/hail/expr/ir/InferPType.scala

```diff
@@ -433,11 +432,153 @@ object InferPType {
           theIR._pType2
         }))
       case In(_, pType: PType) => pType
-      case _: ReadPartition | _: MakeStream => throw new Exception("Node not supported")
+      case NDArrayWrite(_, _) => PVoid
+      case _: InitOp => PVoid
+      case _: SeqOp => PVoid
+      case _: CombOp => PVoid
+      case x@ResultOp(_, _) =>
+        PType.canonical(InferType(x))
+      case _: SerializeAggs => PVoid
+      case _: DeserializeAggs => PVoid
+      case _: Begin => PVoid
+      case _: TableWrite => PVoid
+      case _: TableMultiWrite => PVoid
+      case _: MatrixWrite => PVoid
+      case _: MatrixMultiWrite => PVoid
+      case _: BlockMatrixWrite => PVoid
+      case _: BlockMatrixMultiWrite => PVoid
+      case TableGetGlobals(child) => PType.canonical(child.typ.globalType)
+      case x@TableCollect(_) =>
+        PType.canonical(InferType(x))
+      case TableToValueApply(child, function) => PType.canonical(function.typ(child.typ))
+      case MatrixToValueApply(child, function) => PType.canonical(function.typ(child.typ))
+      case BlockMatrixToValueApply(child, function) =>
+        PType.canonical(function.typ(child.typ))
+      case CollectDistributedArray(contextsIR, globalsIR, contextsName, globalsName, bodyIR) => {
```

Comment:
Attempted to track use of requiredeness on the parent types to decide whether requiredeness on parent was force (as in MakeArray), or inferred from child IR's, but probably got some wrong. Could spend more time looking at it, but getting tired, and I think you'll have fast answers, so may as well post this for review.